### PR TITLE
added edge-pruniing

### PR DESCRIPTION
**Pull request by nishaq503**
added edge-pruning and ignoring extra-vacuous edges.

### CODE REVIEW

File: pyclam/manifold.py [hunk 1]
```diff
@@ -107,6 +107,11 @@ def depth(self) -> int:
         """ The depth in the tree at which the cluster exists. """
         return len(self.name)
 
+    @property
+    def parent(self) -> 'Cluster':
+        """ Returns the parent of self """
+        return self.manifold.select(self.name[:-1])
```
Comment (thoward27) [inline pyclam/manifold.py]: Should we just store a parent reference?

File: pyclam/manifold.py [hunk 2]
```diff
@@ -440,12 +444,40 @@ def _build_edges_matrix(self) -> None:
 
                 batch_left, batch_right = tuple(map(list, np.where(batch_differences <= 0.)))
                 [clusters[l_ + i].neighbors.update({clusters[r_]: batch_distances[l_, r_]}) for l_, r_ in zip(batch_left, batch_right) if (l_ + i) != r_]
+        return
+
+    def _build_edges_pruned(self) -> None:
```
Comment (thoward27) [inline pyclam/manifold.py]: Does this pass unit tests that use it?

File: pyclam/manifold.py [hunk 3]
```diff
@@ -440,12 +444,40 @@ def _build_edges_matrix(self) -> None:
...
+        root = self.manifold.select('')
+
+        for cluster in clusters:
+            ancestor, radius = cluster, cluster.radius
+            while radius <= 0.:
+                ancestor, radius = ancestor.parent, ancestor.parent.radius
+
+            potential_neighbors: Dict['Cluster', Radius] = root.tree_search(
+                point=cluster.medoid,
+                radius=radius * 2 * np.sqrt(2),
```
Comment (thoward27) [inline pyclam/manifold.py]: Weird not seeing dimensionality used here

File: pyclam/manifold.py (hunk 1 above)
Comment (nishaq503) [reply pyclam/manifold.py]: We are not using this anymore. I took it out

File: pyclam/manifold.py (hunk 3 above)
Comment (nishaq503) [reply pyclam/manifold.py]: complexities cancel away

File: pyclam/manifold.py (hunk 2 above)
Comment (nishaq503) [reply pyclam/manifold.py]: It does now

File: pyclam/manifold.py [hunk 4]
```diff
@@ -715,24 +716,26 @@ def _partition_threaded(self, criterion):
 
         return clusters
 
-    def select(self, name: str) -> Cluster:
```
Comment (thoward27) [inline pyclam/manifold.py]: need a docstring here

File: pyclam/datasets.py [hunk 5]
```diff
@@ -30,7 +30,7 @@ def ring_data(n: int, radius: float, noise: float) -> np.ndarray:
     return np.asarray(ring, dtype=np.float64)
 
 
-def bullseye(n: int = 1_000, num_rings: int = 3, noise: float = 0.05) -> Tuple[np.ndarray, List[int]]:
+def bullseye(n: int = 2_000, num_rings: int = 3, noise: float = 0.05) -> Tuple[np.ndarray, List[int]]:
```
Comment (thoward27) [inline pyclam/datasets.py]: dataset changes shouldn't be part of this PR, not that it really matters, but try and keep them singular in purpose

File: pyclam/manifold.py [hunk 6]
```diff
@@ -107,6 +107,9 @@ def depth(self) -> int:
         """ The depth in the tree at which the cluster exists. """
         return len(self.name)
 
+    def distance_from(self, x1: Union[List[int], Data]) -> np.ndarray:
```
Comment (thoward27) [inline pyclam/manifold.py]: Clever, I like this one.

File: pyclam/manifold.py [hunk 7]
```diff
@@ -280,6 +283,7 @@ def partition(self, *criterion) -> Iterable['Cluster']:
         2 children are produced if the cluster can be split, otherwise 1 child is produced.
         """
         if not all((
+                True if self.depth == 0 else self.name[-1] != '0',
```
Comment (thoward27) [inline pyclam/manifold.py]: this reads weirdly, can it be done in two separate boolean clauses

File: pyclam/manifold.py [hunk 8]
```diff
@@ -369,6 +375,7 @@ def __init__(self, *clusters):
 
         # self.clusters is a dictionary of the clusters in the graph and the connected component subgraph that the cluster belongs to.
         self.clusters: Dict[Cluster: 'Graph'] = {c: None for c in clusters}
+        self.count = 0
```
Comment (thoward27) [inline pyclam/manifold.py]: What does this do

File: pyclam/manifold.py [hunk 9]
```diff
@@ -410,42 +417,30 @@ def depth(self) -> int:
     def metric(self) -> Metric:
         return next(iter(self.clusters.keys())).metric
 
-    def _build_edges_matrix(self) -> None:
```
Comment (thoward27) [inline pyclam/manifold.py]: love seeing this go

File: pyclam/manifold.py [hunk 10]
```diff
@@ -410,42 +417,30 @@ def depth(self) -> int:
...
-
-                batch_distances = self.distance(batch_argcenters, argcenters)
-                batch_differences = ((batch_distances - radii).T - batch_radii).T
-
-                batch_left, batch_right = tuple(map(list, np.where(batch_differences <= 0.)))
-                [clusters[l_ + i].neighbors.update({clusters[r_]: batch_distances[l_, r_]}) for l_, r_ in zip(batch_left, batch_right) if (l_ + i) != r_]
-
-        return
-
     def build_edges(self) -> None:
```
Comment (thoward27) [inline pyclam/manifold.py]: Is this the function that is only called at the bottom now? That should be specified if so

Comment (thoward27) [reply pyclam/manifold.py]: This calls top-down, but I thought we discussed that the new function builds bottom-up, can you clarify?

File: pyclam/manifold.py [hunk 11]
```diff
@@ -672,7 +673,7 @@ def build(self, *criterion) -> 'Manifold':
     def build_tree(self, *criterion) -> 'Manifold':
         """ Builds the Cluster-tree. """
         while True:
-            logging.info(f'current depth: {len(self.graphs) - 1}, {len(self.graphs[-1].clusters)} clusters')
+            logging.info(f'depth: {len(self.graphs) - 1}, {len(self.graphs[-1].clusters)} clusters')
```
Comment (thoward27) [inline pyclam/manifold.py]: self.distance?

File: pyclam/manifold.py [hunk 12]
```diff
@@ -715,24 +716,26 @@ def _partition_threaded(self, criterion):
...
-        cluster: Cluster = next(iter(self.graphs[0]))
+        root: Cluster = next(iter(self.graphs[0]))
+        lineage = [root]
         for depth in range(len(name) + 1):
-            partial_name = name[:depth]
-            for child in cluster.children:
-                if child.name == partial_name:
-                    cluster = child
+            for child in lineage[-1].children:
+                if child.name == name[:depth]:
```
Comment (thoward27) [inline pyclam/manifold.py]: Should we make children of a cluster a set keyed on name?

File: pyclam/datasets.py (hunk 5 above)
Comment (nishaq503) [reply pyclam/datasets.py]: I kinda needed to to be able to measure any speedups

File: pyclam/manifold.py (hunk 7 above)
Comment (nishaq503) [reply pyclam/manifold.py]: Checking just self.name[-1] == '0' breaks at the root level.
I would rather just punt this until we find what is really wrong with partition that it produced the empty list that I check for in line 329

File: pyclam/manifold.py (hunk 8 above)
Comment (nishaq503) [reply pyclam/manifold.py]: ugh I've gotten rid of this line so many times

File: pyclam/manifold.py (hunk 10 above)
Comment (nishaq503) [reply pyclam/manifold.py]: It gets called once from the bottom layer. This should indeed be specified.
It iterates over the graphs from the graph at depth 1 to itself. This is handled in in lines 437-442.

Interestingly, the helper _find_neighbors is constructed in such a way so as not to repeat any distance computations in layers over which the graph was already build so it does not even matter if Graph.build_edges gets called multiple times per layer. It will only do the incremental work necessary to complete the building of any graphs.

File: pyclam/manifold.py (hunk 11 above)
Comment (nishaq503) [reply pyclam/manifold.py]: I don't get this

File: pyclam/manifold.py (hunk 4 above)
Comment (nishaq503) [reply pyclam/manifold.py]: On it.

File: pyclam/manifold.py (hunk 12 above)
Comment (nishaq503) [reply pyclam/manifold.py]: There are only one or two children so I do not see the change being worth it

File: pyclam/manifold.py (hunk 6 above)
Comment (nishaq503) [reply pyclam/manifold.py]: It was such common usage

File: pyclam/manifold.py (hunk 7 above)
Comment (nishaq503) [reply pyclam/manifold.py]: This is kinda equivalent to keeping a self.is_leaf flag
